In [129]:
import numpy as np
from scipy import stats
from scipy import integrate

### 1. Number of simulation draws

Suppose the scalar variable θ is approximately normally distributed in a posterior distribution that is summarized by n independent simulation draws. How large does n have to be so that the 2.5% and 97.5% quantiles of θ are specified to an accuracy of 0.1 sd(θ|y)?

(a) Figure this out mathematically, without using simulation.

(b) Check your answer using simulation and show your results.

We know that posterior probabilities are estimated to a standard deviation of 

$\sqrt{p(1-p)/S},$

so that the number of draws $S$ required to estimate a probability to $d$ standard deviations, will be

$S \sim p(1-p)/(d*dens)^2,$

where $dens$ is the associated probability density.

Hence for a normal distribution, in order to estimate the 2.5% quantile (and the 97.5% quantile), we will need:

$S \sim 0.025\times0.975/(0.1 \times dens@2.5\%)^2,$

where $dens@2.5\%$ is the probability density of a Normal distribution at the 2.5% quantile, which is the probability density for a standard normal distribution at $x = -1.96$, which is:

In [71]:
stats.norm.pdf(-1.96)

0.058440944333451476

Hence the number of draws required is:

In [72]:
(0.025 * 0.975)/(0.1 * 0.058441)**2.

713.6895652613404

We can verify this result by simulation: We take 714 draws from a standard normal distribution 2,000 times, and compute the 2.5% and 97.5% quantile in each case.  We then find the sample standard deviation for these, which should be close to 0.10.

In [75]:
true_0_025 = stats.norm.ppf(0.025)
true_0_975 = stats.norm.ppf(0.975)

In [76]:
print(true_0_025, true_0_975)

-1.95996398454 1.95996398454


In [77]:
S = 714
sims = []
for ii in range(2000):
    sims.append(np.random.normal(size=S))

In [83]:
lower = round(714 * 0.025)
upper = round(714 * 0.975)

In [89]:
lowers = []; uppers = []
for ii in range(2000):
    lowers.append(np.sort(sims[ii])[lower])
    uppers.append(np.sort(sims[ii])[upper])

In [99]:
lower_sd = np.sqrt(((lowers - true_0_025)**2).sum()/(len(lowers) - 1))
upper_sd = np.sqrt(((uppers - true_0_975)**2).sum()/(len(uppers) - 1))

In [100]:
print("Standard Deviation of lower estimate: {0:.3f}".format(lower_sd))
print("Standard Deviation of upper estimate: {0:.3f}".format(upper_sd))

Standard Deviation of lower estimate: 0.101
Standard Deviation of upper estimate: 0.104


Similarly, if we want to estimate the median of a normal distribution to 0.02 standard deviations, we will need this many draws:

In [103]:
(0.5*0.5)/(0.02 * stats.norm.pdf(0.))**2

3926.9908169872415

We can again simulate this number of draws 2000 times:

In [107]:
n = 3927; sims = []; mid = round(n/2)

In [110]:
medians = []
for ii in range(2000):
    sims.append(np.random.normal(size=n))
    medians.append(np.sort(sims[ii])[mid])

In [115]:
std = np.sqrt(((np.array(medians) - 0.0)**2).sum()/(len(medians) - 1))

In [117]:
print("Standard deviation for median estimate: {0:.3f}".format(std))

Standard deviation for median estimate: 0.020


### 2. Number of simulation draws
Suppose you are interested in inference for the parameter θ1 in a multivariate posterior distribution, p(θ|y). You draw 100 independent values θ from the posterior distribution of θ and find that the posterior density for θ1 is approximately normal with mean of about 8 and standard deviation of about 4.

(a) Using the average of the 100 draws of θ1 to estimate the posterior mean, E(θ1|y), what is the approximate standard deviation due to simulation variability?

(b) About how many simulation draws would you need to reduce the simulation standard deviation of the posterior mean to 0.1 (thus justifying the presentation of results to one decimal place)?

(c) A more usual summary of the posterior distribution of θ1 is a 95% central posterior interval. Based on the data from 100 draws, what are the approximate simulation standard deviations of the estimated 2.5% and 97.5% quantiles of the posterior distri- bution? (Recall that the posterior density is approximately normal.)

(d) About how many simulation draws would you need to reduce the simulation standard deviations of the 2.5% and 97.5% quantiles to 0.1?

(e) In the eight-schools example of Section 5.5, we simulated 200 posterior draws. What are the approximate simulation standard deviations of the 2.5% and 97.5% quantiles for school A in Table 5.3?

(f) Why was it not necessary, in practice, to simulate more than 200 draws for the SAT coaching example?

(a)

Since

$S \sim p(1-p)/(d\times dens)^2,$

$d \times dens \sim \sqrt{(p(1-p))/S},$

i.e.,

$dens@mean \times d \sim \sqrt{(0.5 \times 0.5)/100} = 0.5/10 = 0.05,$

where $dens@mean$ is the probability density of a normal distribution at the mean, i.e.,:

In [118]:
stats.norm.pdf(0.)

0.3989422804014327

So

$0.3989 \times d \sim 0.05,$

which gives $d$ approx. equal to:

In [121]:
print("Standard Deviation due to simulation variability: {0:.3f}".format(0.05 / 0.3989))

Standard Deviation due to simulation variability: 0.125


(b)

To reduce the standad deviation of the mean to 0.1 (and justify quoting it to one decimal place), we would need:

In [122]:
(0.5*0.5)/(0.1 * stats.norm.pdf(0.))**2

157.07963267948961

around 157 simulation draws.

(c)

Considering the 2.5% and 97.5% percentiles, the standard deviations after 100 draws will be approximately:

$dens@2.5\% \times d \sim \sqrt{(0.025 \times 0.975)/100},$

where $dens@2.5\%$ is:

In [123]:
stats.norm.pdf(-1.96)

0.058440944333451476

so that

$d \sim \sqrt{(0.025 \times 0.975)}/10 / 0.0584$

In [124]:
np.sqrt(0.025 * 0.975) / 10. / 0.0584

0.2673372430821232

(d)

To reduce the standard deviation for the 2.5% and 97.5% quantiles to 0.1, we would need approximately:

In [125]:
(0.025*0.975)/(0.1 * stats.norm.pdf(-1.96))**2

713.69092487844409

(e)

    School Posterior quantiles
           2.5% 25% median 75% 97.5%
       A   −2    7    10   16    31
       
We treat the distribution as normal, hence the 200 simulation draws used to produce the 2.5% and 97.5% quantiles above have simulation standard deviations of approximately:

In [126]:
np.sqrt(0.025 * 0.975 / 200) / 0.0584

0.18903597744708575

times the standard deviation of the distribution, which we can estimate from the 2.5% and 97.5% quantiles themselves.

In [130]:
integrate.dblquad?